### ISSUE TITLE
cache key + other suggestions

### ISSUE THREAD

**Issue by mustafa0x** [henrygd/social-image-server#1]
Hi, thanks for this! A few small requests, if possible.

1. Accept a url param that will be used as a cache key. Users can then use a timestamp, version, or hash of the page to make sure og:images are always fresh. e.g., `<meta property="og:image" content="https://social-image-server.example.com/get?url=yourwebsite.com/blogs/article-1?_cache_key=foobar">`
2. Distinguish requests made to the site, either by query param, or header. This is so the site can customize style and content of the og:image. e.g.,
    - **Custom Header**: "X-OG-Image-Request"
    - **Query Parameter**: "?generate_og_image=true"
3. Persist Chrome, instead of launching a new instance for every screenshot.
4. A nice to have: an endpoint that accepts a list of urls to preemptively generate og:images for. This endpoint would have to be auth-walled (eg, via the `Authorization: Bearer <key>` header).

**Comment by mustafa0x** [comment]
5. Manage domains via (auth-walled) endpoint, or cli command. Changes to the `ALLOWED_DOMAINS` env var requires restarting the service, and hand-editing the compose file (or using an env file). It also makes upgrading the service require manual intervention.

**Comment by henrygd** [comment]
Sorry for the long reply

**1. Accept a url param that will be used as a cache key:**

The problem with this functionality is that it's an easy DoS attack. Someone can just request a million fresh images with different cache keys.

That's why image.social doesn't provide this either -- you have to log into your account to clear a URL.

We do have the `regen` param that always regenerates the image for a given URL. It's useful for finding the right width for a site, or tweaking the delay. Or you could add an internal event hook in your app to regenerate a route if content is updated. But it's not a good idea to expose publicly because of the attack risk.

One workaround is set the `CACHE_TIME` down to 1 hour or something.

We could also add a `/clear` route to clear the cache for a domain or URL.

Another potential way to minimize DoS risk is to add rate limiting, but I'm not sure how the legitimate bots behave and I don't want to block them accidentally.

**2. Distinguish requests made to the site so the site can customize style and content**

Good idea. Query parameter should work for static sites as well so that seems like the better method, right?

**3. Persist Chrome, instead of launching a new instance for every screenshot.**

It is possible currently with `REMOTE_URL`, but it requires you to launch chrome separately from the server. I'll look into this further to see if I can add an option for this and control it within the program.

There are benefits to both approaches depending on your use case. If you don't have a lot of images to generate you may not want chrome always taking up memory.

I think the `chromedp/headless-shell` docker image does persist the browser instance by default, but I need to check that to be sure.

If you use your native chrome installation, I have an example in the readme of how to run it with the necessary flag. I need to update it to include all the other recommended flags, but in theory you could automate it with systemd.

If you run the server in docker but want to use your native chrome install instead of a chrome image, you should be able to do this by giving the docker container access to your host ports. I need to test this. 

**4. Endpoint that accepts a list of urls to preemptively generate og:images**

What's the benefit of doing this versus generating on the fly when the first request comes in? It's also possible to do currently with a loop and the `regen` param.

**5. Manage allowed domains via an (auth-walled) endpoint, or a cli command**

I think this is a good idea in theory. We could add a `/domains` route. `GET` returns active allowed domains. `POST` adds new domains. `DELETE` removes domains.

We can't do both this and the env var, though. It would need to be one or the other.

Pros:

- Adds ability to automate changes to your domain whitelist.

Cons:

- More complex setup process, especially if we add authentication.

- Makes the app less portable since you'd have to either migrate the database / images or re-add the domains. A nice thing about environment variables is that you can spin up your compose file or systemd service anywhere and it will work the same.

- Breaks the current functionality. People using it may not realize that the change was made. Though I'm not completely against breaking changes this early.

I think I lean more towards keeping it as an environment variable. Restarts are nearly instant and I value the portability and ease of setup. Let me know if I'm missing any other pros / cons.

**Comment by mustafa0x** [comment]
Thanks for the detailed response!

> 1. The problem with this functionality is that it's an easy DoS attack.

A simple solution: verify that the page carries the same cache key. This doesn't even include an extra request, as the service has to fetch the page in any case.

**Comment by mustafa0x** [comment]
> 2. Query parameter should work for static sites as well so that seems like the better method, right?

Yeah, if it the user will use JavaScript to customize the markup/style, then a query param is a must.

> 4. What's the benefit of doing this versus generating on the fly when the first request comes in?

No big difference. Just a nice to have. Perhaps when users want a little extra reliability and speed. Or maybe if the user doesn't want to keep Chrome running, so they generate the images, then kill Chrome.

> 5. ...

Another con of env vars: The user will need to modify the compose file to either specify a .env file, or add an env var directly. This'll make automatically updating the compose file (ie via git pull) infeasible.

An env var works for now, but with time I think a cli cmd or endpoint will make more sense. Especially for users that have 10s of domains. Editing long env vars is unwieldy.

**Comment by henrygd** [comment]
> verify that the page carries the same cache key

That makes sense, I'll add it!

**Comment by henrygd** [comment]
Released [v0.0.4](https://github.com/henrygd/social-image-server/releases/tag/v0.0.4) which should take care of 1, 2, and 3, along with other improvements. Let me know if you find any bugs.

Regarding 4 and 5, maybe these can be done with cli commands as you suggested.

Preloading images:

```sh
# binary
social-image-server --preload "example.com/page1,example.com/page2"

# docker
docker run henrygd/social-image-server --preload "example.com/page1,example.com/page2"
```

Managing whitelisted domains:

```sh
# add domain(s) -> returns list of domains
social-image-server --domains add "example.com,example.org"

# delete domains(s) -> returns list of domains 
social-image-server --domains delete "example.com,example.org"

# list domains
social-image-server --domains list
```

I think we can leave the `ALLOWED_DOMAINS` env var in place. If it exists, it pulls the domains form there. If it doesn't, it's pulled from the database and managed by the cli.

We could also add a `purge` command to purge a domain or URL's cache.

Would that work for your case?
